### PR TITLE
feat: implement the interface for room-searching (RoomSearchImpl)

### PR DESCRIPTION
**Pull request by lus-terry**
This interface now contains methods that return a list of suggested rooms depending on what the user is searching for in the search box.

When a user types something in a  [search box,](https://github.com/JakovTomasic/RitehMaps/blob/room_search/client/src/components/Search.tsx) methods inside [RoomSearchImpl](https://github.com/JakovTomasic/RitehMaps/blob/room_search/client/src/logic/impl/RoomSearchImpl.ts) are called.

### CODE REVIEW

File: client/src/logic/impl/RoomSearchImpl.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        
+        // Mapping data into SearchNodeSuggestion type objects
+       const unsortedSuggestions : SearchNodeSuggestion[] = rooms.flatMap((room) => {
+            const { names } = room;
+            const destinationFilter = null;
+            return names.map((name) => new SearchNodeSuggestion(name, destinationFilter));
+        });
+
+        const filteredSuggestions: SearchNodeSuggestion[] = unsortedSuggestions.filter((suggestion) =>
+            suggestion.roomName.toLowerCase().startsWith(searchedText.toLowerCase()) || suggestion.roomName.toLowerCase().includes(searchedText.toLowerCase())
```
Comment (JakovTomasic) [inline client/src/logic/impl/RoomSearchImpl.ts:19]: `suggestion.roomName.toLowerCase().startsWith(searchedText.toLowerCase())` is redundant. The `includes` case already covers a case when a string starts with a substring. Remove the `startsWith` part and leave only the `includes` part of this condition.

File: client/src/components/Search.tsx [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
-  ]);
+  const [dropdownOptions, setDropdownOptions] = useState([]);
+
+  const searchRef = useRef(null);
+
+  const handleInputChange = async (event) => {
+    const inputValue = event.target.value;
+    setInputValue(inputValue);
+
+    const roomSearch = new RoomSearchImpl();
```
Comment (JakovTomasic) [inline client/src/components/Search.tsx:15]: A component should never create a specific implementation of the interface. It should just receive an interface as a prop (dependency injection).

That is a bit harder to do and I wasn't sure if it was possible so I tried and conformed it in #43.

In that pull request, you can see that only a root page (`Home` in your example) can create an implementation of an interface and all other components should receive it as a prop. That is more clearly described in the `React components` section in [the wiki](https://github.com/JakovTomasic/RitehMaps/wiki/Writing-code).

The advantages of this approach are:
- it makes later mocking and changing implementations really easy
- we can easily use different functions for each Search component. E.g. for start and destination searches.
- it follows the "Dependency Inversion" principle of the SOLID principles (if you are curious, I like [this](https://www.baeldung.com/solid-principles#d) explanation of the principle)

**Review by JakovTomasic — CHANGES_REQUESTED** [review]
Great work! The UI works nicely and it's really smooth. I added comments below as suggestions for what can be improved.

I noticed you only display room data and not the professors in the search. After conforming to all of the comments in this review your next task, should you choose to accept it, would be to add professors to the search, too.

Here follows the more precise specification of the needed change.

You should first find an office for each professor using the office ids. If there is no office with the given id don't show the professor (ignore them).
An example from your JSON files would be that "prof. dr. sc. Miroslav Joler" should have an office in officeId = 3. That's room "1-48b".

Notice that SearchNodeSuggestion has the value "name" that will show in the UI. Instead of passing just a professor's name, set it to "professor's name (room name)" where the room name is the first name of the room from the list (in case there are multiple names just pick the first one).
For the above example, the room suggestion in UI should look like "prof. dr. sc. Miroslav Joler (1-48b)".

This suggestion should be visible if the user types e.g. "Jol" or e.g. "48". It shouldn't be visible if the user types e.g. "1-49" or "Ivo".

So, in conclusion, you just need to add a new list (of professors) with a name like described and join it with the current `unsortedSuggestions`. Then filter and sort the same way as you do now.

You can continue with the changes in this branch as it is closely related.

Reminder - first conform all reviewed comments and the pull request, then do the next task.

**Comment by lus-terry** [comment]
I merged the branch 'room_search_proposal' into room_search. 
Also, I added a new feature (a1874cecc6bbb5a9febe75fc42aeb95eb9db626d): users can now also search by professor's name. As you proposed, suggestions are formatted like "prof. dr. sc. Miroslav Joler (1-48b)".
